### PR TITLE
Improve xrtdServer

### PR DESCRIPTION
**Pull request by mrceyhun**
@yuyiguo 

I implemented couple of improvements only for `xrtdTracer`. If it works as expected we can apply it to others: swpop and fwjr.
Sorry if I missed anything. I wrote some suggestions also. Could you please review and try.

Valentin is on vacation. Hope he can review when he come back, fyi @vkuznet 

Improvements are:
- `subscribe` function is moved from `endpointsConfig` to `utils`
- I created `subscribeTest` function which connect directly to DNS of AMQ, instead of one broker like in lb-stomp. It is only for listening topics, I mean subscription in AMQ terminology. We can get review of Valentin in this function.
- I created `safeSubscribe` which tries connecting&subscribing `Config.MaxSubTrial` times. You need to set `MaxSubTrial` in Config file[1]
- `safeUnsubscribe` unsubscribe and disconnect safely.

When we come to `xrtdServer`: 
- IMO you should not subscribe to AMQ inside the for loop. Inside the for loop, you subscribe to AMQ each time you got new message (`case msg := <-sub.C`). 
- I put subscribe outside of for loop.
- I add `defer safeUnsubscribe`, so when there is problem in for loop, `xrtdServer` will call it and then exit.
- `recoverPanic` function is to run `xrtdServer` forever. When there is a panic in `xrtdServer`, because of `defer recoverPanic` line, go will run `recoverPanic` before exit. Because of the fact that `recoverPanic` calls `xrtdServer` again, xrtdServer` will start and continue to run even there is a panic.
- I added `osSigChan` channel to catch `SIGINT` os signals. So, when there is a SIGINT os signal (like ctrl-c keyboard interruption), function will exit.

Suggestions:
- I think you can rename `endpointConsumer` to `consumerTopic` and `endpointProducer` to `producerTopic`. Endpoint is so confusing name :)
- You don't need [default](https://github.com/dmwm/rucio-tracers/blob/main/stompserver/xrootdTracer.go#L284) part in for loop. Because, you're setting `stompRecvTimeout` in stomp connection, see [2] in lb-stomp and [3] in test function. You don't have to worry about it.
- I suggest to increase `stompRecvTimeout` to 10 minutes(600000 millisecond) or so.`5000` millisecond is not a desired number for read timeout of a topic IMO.
- I set `MaxSubTrial` to 10, but generally connection is done in first trial.

[1] 
```json
{
    "stompPassword": "",
    "stompLogin": "", 
    "endpointConsumer": "/topic/xrootd.cms.aaa.ng",
    "endpointProducer": "/topic/cms.rucio.tracer",
    "stompURIProducer": "testxxx:61313",
    "stompURIConsumer": "xxx:61313",
    "interval": 1,
    "verbose": 4,
    "Protocol": "tcp",
    "port": 8888,
    "stompSendTimeout": 5000,
    "stompRecvTimeout": 600000,
    "MaxSubTrial": 10
}
```

[2] https://github.com/vkuznet/lb-stomp/blob/master/stomp.go#L142
[3] https://github.com/mrceyhun/rucio-tracers/blob/xrtd-fix/stompserver/utils.go#L137

### CODE REVIEW

**Comment by yuyiguo** [comment]
Hi Ceyhun,

I will review the code. Here are some quick comments/questions.
_"IMO you should not subscribe to AMQ inside the for loop. Inside the for loop, you subscribe to AMQ each time you got new message (case msg := <-sub.C)."_

Subscription is outside the for loop. I only put it in side the loop in the test to show you even I continued to subscribe on every loop. It could only reach two of the six brokers.

_I think you can rename endpointConsumer to consumerTopic and endpointProducer to producerTopic. Endpoint is so confusing name :)_

Reading the AMQ doc, I understood there were topics and queues. Consumer can consume a topic or a queue with different config. So I would not limit it to topic.

Did you run the code to receive data?

Thanks,
Yuyi

**Comment by mrceyhun** [comment]
@yuyiguo Yes I run and get data. You need to add `"MaxSubTrial": 10` value to config file only.

**Comment by vkuznet** [comment]
@mrceyhun , @yuyiguo before reviewing the code could you please both learn how to use [`gofmt` tool](https://blog.golang.org/gofmt) and properly format your code. This will eliminate different offsets your editors apply to the code and make it common across platforms and editors. Better, try to integrate gofmt in your editor/IDE, see https://golangexample.com/go-development-plugin-for-vim/ or https://github.com/fatih/vim-go. This will help you tremendously and simplify your Go development.

**Comment by mrceyhun** [comment]
@vkuznet I use GoLand IDE and apply go-fmt every file I work. Sorry for the conflict.

**Comment by yuyiguo** [comment]
@vkuznet, I use VS Code. It automatically complies with Golang .

**Comment by yuyiguo** [comment]
Hi @mrceyhun,  @vkuznet 

A few comments/questions:
1. subscribeTest is better to move to lb-stomp in my opinions. 
2. I did not see how the code goes over all the six brokers to get the data. The xrootd topic seemed moved from one broker to another. I saw it in two brokers last week, but this week it was only in one. 
3. the utils is used for some functions that Go does not provide as a language. I would like keep all the stomp releated functions in endpointsConfig. I may  move around stuff later.

**Comment by vkuznet** [comment]
@yuyiguo , I have no idea what VS code is doing, but it seems to me that it does not format the code properly and we see a difference in github commits. Please use `gofmt` to properly format your code before git commits.

Now, regarding the changes:
1. subscribeTest does not belong to lb-stomp library, the latter only provides access to brokers, what you do with them is up to a client
2. I already wrote before that instead of getting the connection from lb-stomp your code should just loop over all connections and use them, e.g.
```
for _, conn := range smgr.Connection {
   // do whatever you want with connection pointer here, e.g. subscribe
}
```
Doing this way you'll get all connections and can subscribe each of them to your endpoint. For instance, you may keep a list or dict of subscriptions to all brokers, and if only one broker provides the data you'll use it. You may generalize code to always use all connections. This will avoid issues you experience if messages only provided by one broker and can move from one broker to another.
3. I don't understand what you're saying here. If you refer to `utils.go` it is just a file which contains function which your code use. You can put them in any file you want and entire codebase will compile the final executable. What I view is that Ceyhun put them in `utils.go` file which seems appropriate.

**Comment by vkuznet** [comment]
I think what Yuyi needs is a list of subscriptions to all brokers (since it is not known which broker will have the messages). As such I suggest the following:
1. add new function to provide list of subscriptions to all brokers, e.g.
```
// subscribeAll is a helper function to subscribe to all StompAMQ brokers
func subscribeAll(endpoint string) ([]*stomp.Subscription, error) {
    var subscriptions []*stomp.Subscription
    smgr := initStomp(endpoint)
    for idx, addr := range smgr.Addresses {
        conn := smgr.ConnectionPool[idx]
        log.Println("stomp connection", conn, addr)
        // subscribe to ActiveMQ topic
        sub, err := conn.Subscribe(endpoint, stomp.AckAuto)
        if err != nil {
            log.Println("unable to subscribe to", endpoint, err)
            return subscriptions, err
        }
        subscriptions = append(subscriptions, sub)
    }
    log.Println("stomp subscriptions", subscriptions)
    return subscriptions, nil
}
```
Please note, the `smgr.Addresses` lists all IP addresses of brokers, while this function will return the list of subscriptions with the same order (if desired you may convert list to a dict).
2. Extend a loop in xrtdServer that it will loop over subscriptions
3. Make new function (put existing for loop from xrtdServer in there) which will take subscription pointer and process a message. The function should take a channel as input and yield message to this channel
4. In xrtdServer make loop to consume messages from a channel used in previous step and process them accordingly.

This plan will allow:
- consume messages from **ALL** brokers, therefore it will avoid guessing which broker to use
- it will allow to reinstantiate subscriptions within individual broker if something will go wrong (xrtdServer subscription will be disconnected)

Can you do that or do you need my help?

**Comment by yuyiguo** [comment]
Hi @vkuznet ,

I did not understand why the load balance library did not connect to different brokers even if it tried to pick a random id/broker?

**Comment by vkuznet** [comment]
@yuyiguo I can't say much without knowing what happen, this PR does not provide any details what actually you experience. But I can say that we tested lb-stomp library (with Lionel) and verified that its logic is correct.

What I think you are mistaken is its use-case. The lb-stomp library should be used when you need evenly spread the load when **sending** data to AMQ. A client will get random broker and send data over. Your use-case is completely opposite. You want to **listen** to **ALL** brokers since you don't know on which broker a message appears. That's why I proposed a solution shown in this thread.

Bottom line, since I was away and didn't pay much attention to previous emails I suggest that you make a summary of the issue and post it somewhere. But from what I understand you should adopt a different strategy when you want to consume data from AMQ, and only use lb-stomp when you send data to AMQ brokers.

**Comment by yuyiguo** [comment]
Hi @vkuznet 
Do you have time today or tomorrow? I want to show you what the problem I am having before I continue to work on either your or Ceyhun's suggestions.

Also VS code has multiple formats as showed on its page [here](https://code.visualstudio.com/docs/languages/go), but not go-fmt.  I just set to my format as :
```
"[go]": {
    "editor.defaultFormatter": "golang.go"
}
```

I hope go-fmt  will also follow golang.go.

**Comment by vkuznet** [comment]
@yuyiguo , feel free to ping me on skype, I'll keep it running today.

Regarding style, VS is a third party product, and `gofmt` is developed by GoLang team and provided as part of GoLang release, therefore it is another way around, the VS should properly adopt what `gofmt` is using and not vice versa. How 3rd party products implemented formatting it is their own business but it seems that it contradicts with a standard provided by `gofmt` tool (in particular number of spaces). Therefore, I still suggest that you run `gofmt` before committing your code. This will ensure that we'll not need to adjust code all the time when we commit changes. On lxplus both `go` and `gofmt` is part of the same package `golang-bin-1.15.13-3.module_el8.4.0+844+08d04f30.x86_64`, and GoLang.org provides it as part of any GoLang release on all platforms (Linux, Windows, MacOS and source code releases).

**Comment by klannon** [comment]
There is a Go extension for VSCode maintained by the Go team at Google, so I assume if you use that, then the formatting should be done correctly.  @yuyiguo, is this what you use?
https://code.visualstudio.com/docs/languages/go
https://code.visualstudio.com/docs/languages/go#_formatting
https://github.com/golang/vscode-go/blob/master/README.md

**Comment by yuyiguo** [comment]
Hi @klannon,

 " Go extension for VSCode " is exact what I am using.

**Comment by klannon** [comment]
Just to be 100% sure, you didn't disable formatting on save in that extension, did you?  In other words, you should ***not*** have the following in your config:
```
"[go]": {
        "editor.formatOnSave": false
}
``` 
Another thing Google seems to suggest is that if you're not careful with the VSCode settings, a general VSCode setting for replacing tabs with spaces can override the "[go]" specific behavior.  You might need something like the following:
```
"[go]": {
   "editor.insertSpaces": false,
}
```
@vkuznet, can you confirm that the issue that you're concerned about is that @yuyiguo's editor is inserting spaces instead of using tabs?

**Comment by vkuznet** [comment]
@klannon , I reported on provided PR which shows different indentation. Therefore it can be either @yuyiguo  or @mrceyhun code, and I'll leave this topic up to them to resolve. What is required is to run `gofmt -l .` in their code areas.

**Comment by yuyiguo** [comment]
@klannon 

Below is my Code/User/settings.json  for go.

```
"[go]": {


        "editor.insertSpaces": false,
        "editor.formatOnSave": true,
        "editor.codeActionsOnSave": {
            "source.organizeImports": true
        },
        "editor.defaultFormatter": "golang.go"
    },

```

**Comment by klannon** [comment]
Sounds reasonable.  I see that @mrceyhun, is also applying the formatting, except in one commit where it was forgotten, so hopefully we can stop discussing formatting and go back to discussing the code?

**Comment by mrceyhun** [comment]
@klannon yes, I missed formatting in one commit and fixed it by force pushing to that commit.
